// @privy-io/node/resources/intents.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Intents = void 0;
const resource_1 = require("../core/resource.js");
const pagination_1 = require("../core/pagination.js");
const headers_1 = require("../internal/headers.js");
const path_1 = require("../internal/utils/path.js");
class Intents extends resource_1.APIResource {
    /**
     * List intents for an app. Returns a paginated list of intents with their current
     * status and details.
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const intentResponse of client.intents.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/v1/intents', (pagination_1.Cursor), { query, ...options });
    }
    /**
     * Create an intent to add a rule to a policy. The intent must be authorized by the
     * policy owner before it can be executed.
     *
     * @example
     * ```ts
     * const ruleIntentResponse =
     *   await client.intents.createPolicyRule('policy_id', {
     *     action: 'ALLOW',
     *     conditions: [
     *       {
     *         field: 'to',
     *         field_source: 'ethereum_transaction',
     *         operator: 'eq',
     *         value: 'string',
     *       },
     *     ],
     *     method: 'eth_sendTransaction',
     *     name: 'x',
     *   });
     * ```
     */
    createPolicyRule(policyID, params, options) {
        const { 'privy-request-expiry': privyRequestExpiry, ...body } = params;
        return this._client.post((0, path_1.path) `/v1/intents/policies/${policyID}/rules`, {
            body,
            ...options,
            headers: (0, headers_1.buildHeaders)([
                { ...(privyRequestExpiry != null ? { 'privy-request-expiry': privyRequestExpiry } : undefined) },
                options?.headers,
            ]),
        });
    }
    /**
     * Create an intent to delete a rule from a policy. The intent must be authorized
     * by the policy owner before it can be executed.
     *
     * @example
     * ```ts
     * const ruleIntentResponse =
     *   await client.intents.deletePolicyRule('rule_id', {
     *     policy_id: 'policy_id',
     *   });
     * ```
     */
    deletePolicyRule(ruleID, params, options) {
        const { policy_id, 'privy-request-expiry': privyRequestExpiry } = params;
        return this._client.delete((0, path_1.path) `/v1/intents/policies/${policy_id}/rules/${ruleID}`, {
            ...options,
            headers: (0, headers_1.buildHeaders)([
                { ...(privyRequestExpiry != null ? { 'privy-request-expiry': privyRequestExpiry } : undefined) },
                options?.headers,
            ]),
        });
    }
    /**
     * Retrieve an intent by ID. Returns the intent details including its current
     * status, authorization details, and execution result if applicable.
     *
     * @example
     * ```ts
     * const intentResponse = await client.intents.get(
     *   'intent_id',
     * );
     * ```
     */
    get(intentID, options) {
        if (intentID === '') {
            throw new Error('intentID must not be an empty string');
        }
        return this._client.get((0, path_1.path) `/v1/intents/${intentID}`, options);
    }
    /**
     * Create an intent to execute an RPC method on a wallet. The intent must be
     * authorized by either the wallet owner or signers before it can be executed.
     *
     * @example
     * ```ts
     * const rpcIntentResponse = await client.intents.rpc(
     *   'wallet_id',
     *   {
     *     method: 'eth_signTransaction',
     *     params: { transaction: {} },
     *   },
     * );
     * ```
     */
    rpc(walletID, params, options) {
        const { 'privy-request-expiry': privyRequestExpiry, ...body } = params;
        return this._client.post((0, path_1.path) `/v1/intents/wallets/${walletID}/rpc`, {
            body,
            ...options,
            headers: (0, headers_1.buildHeaders)([
                { ...(privyRequestExpiry != null ? { 'privy-request-expiry': privyRequestExpiry } : undefined) },
                options?.headers,
            ]),
        });
    }
    /**
     * Create an intent to update a key quorum. The intent must be authorized by the
     * key quorum members before it can be executed.
     *
     * @example
     * ```ts
     * const keyQuorumIntentResponse =
     *   await client.intents.updateKeyQuorum('key_quorum_id');
     * ```
     */
    updateKeyQuorum(keyQuorumID, params, options) {
        const { 'privy-request-expiry': privyRequestExpiry, ...body } = params;
        return this._client.patch((0, path_1.path) `/v1/intents/key_quorums/${keyQuorumID}`, {
            body,
            ...options,
            headers: (0, headers_1.buildHeaders)([
                { ...(privyRequestExpiry != null ? { 'privy-request-expiry': privyRequestExpiry } : undefined) },
                options?.headers,
            ]),
        });
    }
    /**
     * Create an intent to update a policy. The intent must be authorized by the policy
     * owner before it can be executed.
     *
     * @example
     * ```ts
     * const policyIntentResponse =
     *   await client.intents.updatePolicy('policy_id');
     * ```
     */
    updatePolicy(policyID, params, options) {
        const { 'privy-request-expiry': privyRequestExpiry, ...body } = params;
        return this._client.patch((0, path_1.path) `/v1/intents/policies/${policyID}`, {
            body,
            ...options,
            headers: (0, headers_1.buildHeaders)([
                { ...(privyRequestExpiry != null ? { 'privy-request-expiry': privyRequestExpiry } : undefined) },
                options?.headers,
            ]),
        });
    }
    /**
     * Create an intent to update a rule on a policy. The intent must be authorized by
     * the policy owner before it can be executed.
     *
     * @example
     * ```ts
     * const ruleIntentResponse =
     *   await client.intents.updatePolicyRule('rule_id', {
     *     policy_id: 'policy_id',
     *     action: 'ALLOW',
     *     conditions: [
     *       {
     *         field: 'to',
     *         field_source: 'ethereum_transaction',
     *         operator: 'eq',
     *         value: 'string',
     *       },
     *     ],
     *     method: 'eth_sendTransaction',
     *     name: 'x',
     *   });
     * ```
     */
    updatePolicyRule(ruleID, params, options) {
        const { policy_id, 'privy-request-expiry': privyRequestExpiry, ...body } = params;
        return this._client.patch((0, path_1.path) `/v1/intents/policies/${policy_id}/rules/${ruleID}`, {
            body,
            ...options,
            headers: (0, headers_1.buildHeaders)([
                { ...(privyRequestExpiry != null ? { 'privy-request-expiry': privyRequestExpiry } : undefined) },
                options?.headers,
            ]),
        });
    }
    /**
     * Create an intent to update a wallet. The intent must be authorized by the wallet
     * owner before it can be executed.
     *
     * @example
     * ```ts
     * const walletIntentResponse =
     *   await client.intents.updateWallet('wallet_id');
     * ```
     */
    updateWallet(walletID, params, options) {
        const { 'privy-request-expiry': privyRequestExpiry, ...body } = params;
        return this._client.patch((0, path_1.path) `/v1/intents/wallets/${walletID}`, {
            body,
            ...options,
            headers: (0, headers_1.buildHeaders)([
                { ...(privyRequestExpiry != null ? { 'privy-request-expiry': privyRequestExpiry } : undefined) },
                options?.headers,
            ]),
        });
    }
}
exports.Intents = Intents;
