// @privy-io/node/public-api/services/utils/auth.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrivyAuthUtils = void 0;
const auth_1 = require("../../../lib/auth.js");
class PrivyAuthUtils {
    constructor(privyApiClient, appJwks) {
        this.appJwks = appJwks;
        this.privyAppID = privyApiClient.appID;
    }
    /**
     * Verifies the access token, and returns the payload if it is valid.
     *
     * @param accessToken - The access token to verify.
     * @returns The payload of the token if it is valid.
     * @throws If the token is invalid.
     */
    async verifyAccessToken(accessToken) {
        return (0, auth_1.verifyAccessToken)({
            access_token: accessToken,
            app_id: this.privyAppID,
            verification_key: this.appJwks,
        });
    }
    /**
     * Verifies the authentication token, and returns the payload if it is valid.
     *
     * @param authToken - The authentication token to verify.
     * @returns The payload of the token if it is valid.
     * @throws If the token is invalid.
     * @deprecated Use `verifyAccessToken` instead.
     */
    async verifyAuthToken(authToken) {
        return this.verifyAccessToken(authToken);
    }
    async verifyIdentityToken(identityToken) {
        return (0, auth_1.verifyIdentityToken)({
            identity_token: identityToken,
            app_id: this.privyAppID,
            verification_key: this.appJwks,
        });
    }
}
exports.PrivyAuthUtils = PrivyAuthUtils;
