// @privy-io/node/public-api/services/apps.mjs
import { Apps } from "../../resources/apps/apps.mjs";
export class PrivyAppsService extends Apps {
    constructor(privyApiClient) {
        super(privyApiClient);
    }
    async getAllowlist() {
        return this.allowlist.list(this._client.appID);
    }
    async inviteToAllowlist(entry) {
        return this.allowlist.create(this._client.appID, entry);
    }
    async removeFromAllowlist(entry) {
        return this.allowlist.delete(this._client.appID, entry);
    }
    async getSettings() {
        return this.get(this._client.appID);
    }
    async getTestAccessToken(params) {
        const { data: testAccounts } = await this.getTestCredentials(this._client.appID);
        if (!testAccounts || testAccounts.length === 0) {
            throw new Error('No test accounts found for this app. Create test accounts in the Privy dashboard.');
        }
        let account;
        if (params && 'email' in params) {
            account = testAccounts.find((a) => a.email === params.email);
            if (!account) {
                throw new Error(`No test account found with email "${params.email}". Available: ${testAccounts
                    .map((a) => a.email)
                    .join(', ')}`);
            }
        }
        else if (params && 'phone_number' in params) {
            account = testAccounts.find((a) => a.phone_number === params.phone_number);
            if (!account) {
                throw new Error(`No test account found with phone number "${params.phone_number}". Available: ${testAccounts
                    .map((a) => a.phone_number)
                    .join(', ')}`);
            }
        }
        else {
            account = testAccounts[0];
        }
        const authResponse = await this._client.post('/v1/passwordless/authenticate', {
            body: { email: account.email, code: account.otp_code },
        });
        if (!authResponse.token) {
            throw new Error('Unable to authenticate with the test account.');
        }
        return { access_token: authResponse.token };
    }
}
