// @privy-io/node/public-api/services/key-quorums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrivyKeyQuorumsService = void 0;
const authorization_1 = require("../../lib/authorization.js");
const resources_1 = require("../../resources.js");
class PrivyKeyQuorumsService extends resources_1.KeyQuorums {
    constructor(privyApiClient, privyClient) {
        super(privyApiClient);
        this.privyClient = privyClient;
    }
    async update(keyQuorumId, { authorization_context: authorizationContext = {}, request_expiry: requestExpiry, ...params }) {
        const { headers } = await (0, authorization_1.prepareRequest)(this.privyClient, this._client.appID, {
            authorizationContext,
            requestExpiry: requestExpiry ?? this.privyClient.getRequestExpiry(),
            method: 'PATCH',
            url: `${this._client.baseURL}/v1/key_quorums/${keyQuorumId}`,
            body: params,
        });
        return await this._update(keyQuorumId, { ...params, ...headers });
    }
    async delete(keyQuorumId, { authorization_context: authorizationContext = {}, request_expiry: requestExpiry, ...params }) {
        const { headers } = await (0, authorization_1.prepareRequest)(this.privyClient, this._client.appID, {
            authorizationContext,
            requestExpiry: requestExpiry ?? this.privyClient.getRequestExpiry(),
            method: 'DELETE',
            url: `${this._client.baseURL}/v1/key_quorums/${keyQuorumId}`,
            body: params,
        });
        return await this._delete(keyQuorumId, { ...params, ...headers });
    }
}
exports.PrivyKeyQuorumsService = PrivyKeyQuorumsService;
