// @privy-io/node/public-api/services/key-quorums.mjs
import { prepareRequest } from "../../lib/authorization.mjs";
import { KeyQuorums, } from "../../resources.mjs";
export class PrivyKeyQuorumsService extends KeyQuorums {
    constructor(privyApiClient, privyClient) {
        super(privyApiClient);
        this.privyClient = privyClient;
    }
    async update(keyQuorumId, { authorization_context: authorizationContext = {}, request_expiry: requestExpiry, ...params }) {
        const { headers } = await prepareRequest(this.privyClient, this._client.appID, {
            authorizationContext,
            requestExpiry: requestExpiry ?? this.privyClient.getRequestExpiry(),
            method: 'PATCH',
            url: `${this._client.baseURL}/v1/key_quorums/${keyQuorumId}`,
            body: params,
        });
        return await this._update(keyQuorumId, { ...params, ...headers });
    }
    async delete(keyQuorumId, { authorization_context: authorizationContext = {}, request_expiry: requestExpiry, ...params }) {
        const { headers } = await prepareRequest(this.privyClient, this._client.appID, {
            authorizationContext,
            requestExpiry: requestExpiry ?? this.privyClient.getRequestExpiry(),
            method: 'DELETE',
            url: `${this._client.baseURL}/v1/key_quorums/${keyQuorumId}`,
            body: params,
        });
        return await this._delete(keyQuorumId, { ...params, ...headers });
    }
}
