// @privy-io/node/public-api/services/policies.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrivyPoliciesService = void 0;
const authorization_1 = require("../../lib/authorization.js");
const resources_1 = require("../../resources.js");
class PrivyPoliciesService extends resources_1.Policies {
    constructor(privyApiClient, privyClient) {
        super(privyApiClient);
        this.privyClient = privyClient;
    }
    create({ idempotency_key: idempotencyKey, ...params }) {
        return super.create({
            ...params,
            ...(idempotencyKey && { 'privy-idempotency-key': idempotencyKey }),
        });
    }
    async update(policyId, { authorization_context: authorizationContext = {}, request_expiry: requestExpiry, ...params }) {
        const { headers } = await (0, authorization_1.prepareRequest)(this.privyClient, this._client.appID, {
            authorizationContext,
            requestExpiry: requestExpiry ?? this.privyClient.getRequestExpiry(),
            method: 'PATCH',
            url: `${this._client.baseURL}/v1/policies/${policyId}`,
            body: params,
        });
        return await this._update(policyId, { ...params, ...headers });
    }
    async delete(policyId, { authorization_context: authorizationContext = {}, request_expiry: requestExpiry, ...params }) {
        const { headers } = await (0, authorization_1.prepareRequest)(this.privyClient, this._client.appID, {
            authorizationContext,
            requestExpiry: requestExpiry ?? this.privyClient.getRequestExpiry(),
            method: 'DELETE',
            url: `${this._client.baseURL}/v1/policies/${policyId}`,
            body: params,
        });
        return await this._delete(policyId, { ...params, ...headers });
    }
    async createRule(policyId, { authorization_context: authorizationContext = {}, request_expiry: requestExpiry, ...params }) {
        const { headers } = await (0, authorization_1.prepareRequest)(this.privyClient, this._client.appID, {
            authorizationContext,
            requestExpiry: requestExpiry ?? this.privyClient.getRequestExpiry(),
            method: 'POST',
            url: `${this._client.baseURL}/v1/policies/${policyId}/rules`,
            body: params,
        });
        return await this._createRule(policyId, { ...params, ...headers });
    }
    async updateRule(ruleId, { authorization_context: authorizationContext = {}, request_expiry: requestExpiry, policy_id: policyId, ...params }) {
        const { headers } = await (0, authorization_1.prepareRequest)(this.privyClient, this._client.appID, {
            authorizationContext,
            requestExpiry: requestExpiry ?? this.privyClient.getRequestExpiry(),
            method: 'PATCH',
            url: `${this._client.baseURL}/v1/policies/${policyId}/rules/${ruleId}`,
            body: params,
        });
        return await this._updateRule(ruleId, { ...params, policy_id: policyId, ...headers });
    }
    async deleteRule(ruleId, { authorization_context: authorizationContext = {}, request_expiry: requestExpiry, policy_id: policyId, ...params }) {
        const { headers } = await (0, authorization_1.prepareRequest)(this.privyClient, this._client.appID, {
            authorizationContext,
            requestExpiry: requestExpiry ?? this.privyClient.getRequestExpiry(),
            method: 'DELETE',
            url: `${this._client.baseURL}/v1/policies/${policyId}/rules/${ruleId}`,
            body: params,
        });
        return await this._deleteRule(ruleId, { ...params, policy_id: policyId, ...headers });
    }
}
exports.PrivyPoliciesService = PrivyPoliciesService;
