// @privy-io/node/lib/cryptography.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateP256KeyPair = generateP256KeyPair;
exports.importPKCS8PrivateKey = importPKCS8PrivateKey;
exports.setupHPKERecipient = setupHPKERecipient;
exports.setupHPKESender = setupHPKESender;
const chacha20poly1305_1 = require("@hpke/chacha20poly1305");
const core_1 = require("@hpke/core");
const nist_1 = require("@noble/curves/nist");
const base64_1 = require("../internal/utils/base64.js");
/**
 * Returns the runtime's `SubtleCrypto` implementation.
 *
 * We rely on `globalThis.crypto.subtle` for broad runtime support (Node.js 20+, Deno, Bun, Workers/Edge).
 *
 * @internal
 */
function getSubtleCrypto() {
    const subtle = globalThis.crypto?.subtle;
    if (!subtle) {
        throw new Error('`crypto.subtle` is not defined as a global; Either run in a runtime that provides WebCrypto, or polyfill `globalThis.crypto`');
    }
    return subtle;
}
/**
 * Generates a P-256 key pair suitable for Privy resource ownership and request
 * authorization signing.
 *
 * @returns A P-256 key pair, in base64-encoded DER format.
 *
 * @example
 * const keypair = await generateP256KeyPair();
 * const wallet = await privy.wallets().create({
 *   chain_type: '...',
 *   owner: { public_key: keypair.publicKey },
 * });
 * const response = await privy.wallets().rawSign(wallet.id, {
 *   params: { hash: '...' },
 *   authorization_context: {
 *     authorization_private_keys: [keypair.privateKey]
 *   },
 * });
 */
async function generateP256KeyPair() {
    const subtle = getSubtleCrypto();
    const keyPair = await subtle.generateKey({ name: 'ECDSA', namedCurve: 'P-256' }, true, ['sign', 'verify']);
    const [publicKeyDer, privateKeyDer] = await Promise.all([
        subtle.exportKey('spki', keyPair.publicKey),
        subtle.exportKey('pkcs8', keyPair.privateKey),
    ]);
    return {
        publicKey: (0, base64_1.toBase64)(new Uint8Array(publicKeyDer)),
        privateKey: (0, base64_1.toBase64)(new Uint8Array(privateKeyDer)),
    };
}
/**
 * Imports a P-256 private key for use with the `@noble/curves` library.
 *
 * @param privateKey - A base64-encoded PKCS8-formatted private key, with no PEM headers.
 * @returns A private key object for the P-256 curve.
 * @internal
 */
function importPKCS8PrivateKey(privateKey) {
    const strippedPrivateKey = privateKey
        .replace(AUTHORIZATION_PRIVATE_KEY_PREFIX, '')
        .replace(WALLET_API_PRIVATE_KEY_PREFIX, '');
    // We fall back to `Buffer` here as Uint8Array.fromBase64 is not widely supported yet
    const pkcs8Bytes = Buffer.from(strippedPrivateKey, 'base64');
    const privateKeyStart = pkcs8Bytes.indexOf(Buffer.from([0x04, 0x20]));
    if (privateKeyStart === -1) {
        throw new Error('Invalid wallet authorization private key');
    }
    const privateKeyBytes = pkcs8Bytes.subarray(privateKeyStart + 2, privateKeyStart + 34);
    return nist_1.p256.Point.Fn.fromBytes(privateKeyBytes);
}
/**
 * Sets up HPKE for securely requesting a payload (as the recipient)
 * from a sender (e.g. the Privy API).
 * @returns The HPKE receiver object.
 * @internal
 */
async function setupHPKERecipient() {
    const suite = new core_1.CipherSuite({
        kem: new core_1.DhkemP256HkdfSha256(),
        kdf: new core_1.HkdfSha256(),
        aead: new chacha20poly1305_1.Chacha20Poly1305(),
    });
    const keypair = await suite.kem.generateKeyPair();
    const subtle = getSubtleCrypto();
    const publicKeySpki = await subtle.exportKey('spki', keypair.publicKey);
    return {
        publicKeySpki: new Uint8Array(publicKeySpki),
        decryptPayload: async (encapsulatedKey, ciphertext) => {
            const recipient = await suite.createRecipientContext({
                recipientKey: keypair.privateKey,
                enc: encapsulatedKey,
            });
            const decodedBytes = await recipient.open(ciphertext);
            return new Uint8Array(decodedBytes);
        },
    };
}
async function setupHPKESender() {
    const suite = new core_1.CipherSuite({
        kem: new core_1.DhkemP256HkdfSha256(),
        kdf: new core_1.HkdfSha256(),
        aead: new chacha20poly1305_1.Chacha20Poly1305(),
    });
    return {
        encryptPayload: async (publicKey, payload) => {
            const recipientPublicKey = await suite.kem.deserializePublicKey(publicKey);
            const sender = await suite.createSenderContext({
                recipientPublicKey,
            });
            const ciphertext = await sender.seal(payload);
            return {
                ciphertext: new Uint8Array(ciphertext),
                encapsulatedKey: new Uint8Array(sender.enc),
            };
        },
    };
}
/** This prefix is no longer used, but we need to support existing keys */
const WALLET_API_PRIVATE_KEY_PREFIX = 'wallet-api:';
const AUTHORIZATION_PRIVATE_KEY_PREFIX = 'wallet-auth:';
